### PR TITLE
[coordinator] Add support for prometheus remote write compatible backend

### PR DESCRIPTION
**Pull request by Antanukas**
**What this PR does / why we need it**:

Adds support for new backend type - prometheus remote write. This enables using M3 Coordinator and M3 Aggregator with Prometheus Remote Write compatible API.

**Special notes for your reviewer**:

Read part is not implemented mainly because using remote read API is not efficient since it requires to load raw timeseries.

In this PR integration with aggregator is not fully implemented. For now one can specify endpoints with no storage policy only so they will receive raw writes.

I plan to fix this in subsequent PR.

Example of configuration when data aggregation will be available:
```
backend: prom-remote

prometheusRemoteBackend:
  endpoints:
    - name: raw
      address: "http://raw-metrics/api/v1/prom/remote/write"
    - name: aggregated_30d
      address: "http://aggregated_30d/api/v1/prom/remote/write"
      storagePolicy:
        retention: 720h
        resolution: 1m
    - name: aggregated_60d
      address: "http://aggregated_60d/api/v1/prom/remote/write"
      storagePolicy:
        retention: 1440h
        resolution: 1m
		
clusterManagement:
  etcd:
    env: test-env
    zone: embedded
    service: m3db
    cacheDir: /var/lib/m3kv
    etcdClusters:
    - zone: embedded
      endpoints:
      - http://etcd-0:2379
      - http://etcd-1:2379
      - http://etcd-2:2379
```
**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
This feature will be available and can be enabled using configuration switch. 
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
Documentation will follow after this PR is merged.
```

### CODE REVIEW

File: src/query/server/query.go
Comment (vpranckaitis) [inline src/query/server/query.go:507]: Not sure if this breaks anything, but I believe `clusterNamespacesWatcher` here is not initialized. At least m3db `case` calls `initClusters()`

https://github.com/m3db/m3/blob/43a0942cf449892eddc04aa004ab63bd7ede3f3c/src/query/server/query.go#L454-L458

Also, what's unusual is that the cleanup of `clusterNamespaceWatcher` is defined in `newM3DBStorage()` function and not where it's initialized (ditto for `clusters` in the snippet below). It might make sense moving the cleanup out of there.

https://github.com/m3db/m3/blob/43a0942cf449892eddc04aa004ab63bd7ede3f3c/src/query/server/query.go#L720-L736

**Review by vpranckaitis — COMMENTED** [review]
A few initial comments, will continue reviewing tomorrow.

File: src/query/storage/promremote/storage_unsupported.go
Comment (linasm) [reply src/query/storage/promremote/storage_unsupported.go:85]: Actually,
```
return fmt.Errorf("promStorage: %s method is not supported", name)
```

File: src/query/server/query.go
Comment (Antanukas) [reply src/query/server/query.go:507]: Yeah this one is confusing. I will need to double check how this namespace watcher is used in downsampler and writer. I might need to register namespaces to it from endpoints. I think its ok if its not initialised since some backends (e.g grpc backend) do not initialise it. But I will move it out of `newM3DBStorage` to avoid similar mistake in the future.

File: src/cmd/services/m3query/config/config.go
Comment (Antanukas) [inline src/cmd/services/m3query/config/config.go:771]: I'm still wondering if I should make these required or optional. 0 kinda means raw unaggregated namespace. But should it be explicit in configuration?

File: src/query/storage/promremote/storage.go
Comment (Antanukas) [inline src/query/storage/promremote/storage.go:117]: I think I better use map by name here instead of address since its possible to configure multiple endpoints with same address

File: src/query/storage/promremote/query_mapper.go
Comment (linasm) [inline src/query/storage/promremote/query_mapper.go:32]: Nit: `convertAndEncodeWriteQuery`? Because it is effectively encoding `prompb.WriteRequest` and not the `storage.WriteQuery`.

File: src/query/storage/promremote/storage.go
Comment (linasm) [inline src/query/storage/promremote/storage.go:90]:
```suggestion
    return multiErr.FinalError()
```

File: src/query/storage/promremote/storage.go
Comment (linasm) [inline src/query/storage/promremote/storage.go:131]: Nit:
```suggestion
	metrics.ReportSuccessOrError(err, time.Since(start))
	if err != nil {
		return err
	}

	defer func() { _ = resp.Body.Close() }()
```

File: src/query/storage/promremote/storage.go
Comment (linasm) [inline src/query/storage/promremote/storage.go:137]: Not sure if we want to expose those errors to the client.

File: src/query/storage/promremote/storage_test.go
Comment (linasm) [inline src/query/storage/promremote/storage_test.go:253]: Nit - you could make it more generic (and perhaps we already have something like this):
```suggestion
func closeWithCheck(t *testing.T, c io.Closer) {
	require.NoError(t, c.Close())
}
```

File: src/query/storage/promremote/storage_test.go
Comment (linasm) [inline src/query/storage/promremote/storage_test.go:112]: Why not resetting `promLongRetention2`?

File: src/query/storage/promremote/storage_test.go
Comment (linasm) [inline src/query/storage/promremote/storage_test.go:65]: Nit: this use of duration for a time looks a bit weird.

File: src/query/storage/promremote/storage_test.go
Comment (linasm) [inline src/query/storage/promremote/storage_test.go:198]: I think this test case can be dropped in favour of `send write to multiple instances configured with same retention`.
I don't see anything specific about this third retention (vs. the other two).

File: src/query/storage/promremote/storage_test.go
Comment (linasm) [inline src/query/storage/promremote/storage_test.go:204]: I don't get this comment and the use of `storagemetadata.AggregatedMetricsType`. Where is the implementation for a check like this?

File: src/query/storage/promremote/storage_unsupported.go
Comment (vpranckaitis) [inline src/query/storage/promremote/storage_unsupported.go:34]: nit: not quite sure about splitting `promStorage` method definitions across two files, people reading code might find it unexpected

**Review by vpranckaitis — APPROVED** [review]
LGTM, though it would be best if someone else would take a look, too.

File: src/query/storage/promremote/storage.go
Comment (linasm) [inline src/query/storage/promremote/storage.go:66]: We need to clarify the requirements to make sure we don't have to return an error in case no matching endpoint was found. Maybe there should be an option to specify some default "catch everything else" endpoint.

File: src/query/storage/promremote/storage_test.go
Comment (linasm) [inline src/query/storage/promremote/storage_test.go:236]: 👍 for this test case.

File: src/query/storage/promremote/storage_unsupported.go
Comment (Antanukas) [reply src/query/storage/promremote/storage_unsupported.go:34]: I wanted to hide these unimplemented methods to make it easier reading the implemented part.

File: src/cmd/services/m3query/config/config.go
Comment (Antanukas) [reply src/cmd/services/m3query/config/config.go:771]: Ill do it. There is same approach with CarbonIngestionStoragePolicyConfiguration

File: src/query/server/query.go
Comment (Antanukas) [reply src/query/server/query.go:361]: It should be harmless to close it multiple times. I will remove close from m3dbstorage close. This `clusterNamespacesWatcher` seems to be needed in order to notify aggregation tier about namespace changes, so when everything is being closed the order when `clusterNamespacesWatcher` should not matter.

File: src/query/server/query.go
Comment (Antanukas) [reply src/query/server/query.go:344]: I think its less error prone like this. This watcher is just an in memory watcher of namespaces. Until now only m3db storage was using and actually adding namespaces to this watcher. It's also passed as parameter to CarbonIngestion which in theory could have been uninitialised before if carbon is used with other backends.

File: src/query/server/query_test.go
Comment (Antanukas) [reply src/query/server/query_test.go:294]: We do, the test bodies are similar and can reuse some of the server startup code I introduced. I will do it in separate PR to not blow up this PR even more.

File: src/query/storage/promremote/storage.go
Comment (Antanukas) [reply src/query/storage/promremote/storage.go:137]: I will add TODO and will fix in another PR where I have logger available.

File: src/query/storage/promremote/storage_test.go
Comment (Antanukas) [reply src/query/storage/promremote/storage_test.go:204]: yeah its a leftover. MetricsType is not actually used when matching endpoints

File: src/query/storage/promremote/storage.go
Comment (Antanukas) [reply src/query/storage/promremote/storage.go:66]: Good idea. Will followup with a PR.

File: src/query/storage/promremote/storage.go
Comment (Antanukas) [reply src/query/storage/promremote/storage.go]: I actually noticed that success was being recorded in case non 200 http status was returned. Fixed that.

File: src/query/server/query_test.go
Comment (linasm) [reply src/query/server/query_test.go:294]: Where? I only see `TestGRPCBackend`, nothing for `M3DBStorageType`.

File: src/query/storage/promremote/doc.go
Comment (linasm) [inline src/query/storage/promremote/doc.go:21]: Why not put this comment into types.go and save 20 lines of legal stuff?

File: src/query/storage/promremote/options.go
Comment (linasm) [inline src/query/storage/promremote/options.go:124]: Perhaps now that `storagePolicy` itself is optional, you should check that resolution/retention are strictly positive?

File: src/query/storage/promremote/query_coverter.go
Comment (linasm) [inline src/query/storage/promremote/query_coverter.go:45]: No no, this one only converts. `convertAndEncodeWriteQuery` should be the one above.

File: src/query/storage/promremote/storage.go
Comment (linasm) [reply src/query/storage/promremote/storage.go:133]: So you can return a generic error now, and TODO just logging.

File: src/query/storage/promremote/storage_unsupported.go
Comment (linasm) [reply src/query/storage/promremote/storage_unsupported.go:34]: I agree with @vpranckaitis, this is unusual (having some methods for the struct in another file). 
Perhaps it would be more clear to do it like this:
```go
type unimplementedPromStorageMethods struct {}

func (p *unimplementedPromStorageMethods) FetchProm(...
```
And then embed this `unimplementedPromStorageMethods` into the actual struct.

File: src/query/server/query_test.go
Comment (Antanukas) [reply src/query/server/query_test.go:294]: there are tests named `TestWriteH1` and `TestWriteH2C` that use configuration for m3db backend. However they mock out m3db client session so not sure how well they test m3db integration. I think it would be a good idea to at least add test case that runs M3DB in memory and writes/reads metric from it.

File: src/query/storage/promremote/query_converter_test.go
Comment (linasm) [inline src/query/storage/promremote/query_converter_test.go:17]: Nit (I find the usage of "Query" for writes ver confusing in general):
```suggestion
func TestWriteQueryConverter(t *testing.T) {
```

File: src/query/storage/promremote/query_converter_test.go
Comment (linasm) [inline src/query/storage/promremote/query_converter_test.go:63]: You only have one test case. What's the point of test table?

File: src/query/storage/promremote/query_coverter.go
Comment (linasm) [inline src/query/storage/promremote/query_coverter.go:35]: Extract an error with fixed message into a global var.

File: src/query/storage/promremote/query_coverter.go
Comment (Antanukas) [reply src/query/storage/promremote/query_coverter.go:45]: facepalm

File: src/query/storage/promremote/storage_unsupported.go
Comment (Antanukas) [reply src/query/storage/promremote/storage_unsupported.go:34]: Ok, ill just put these unimplemented methods at the end of the source file after a comment.

Comment (Antanukas) [reply src/query/storage/promremote/storage_unsupported.go]: introduced struct, looks pretty neat actually.

File: src/query/storage/promremote/query_converter_test.go
Comment (Antanukas) [reply src/query/storage/promremote/query_converter_test.go]: added more cases but still removed table

File: src/query/storage/promremote/storage.go
Comment (linasm) [inline src/query/storage/promremote/storage.go:137]: Make this a global constant to avoid allocations when things go wrong.

File: src/query/storage/promremote/query_converter_test.go
Comment (linasm) [inline src/query/storage/promremote/query_converter_test.go:56]: I think you should resurrect the test table if you have multiple test cases now.

File: src/query/storage/promremote/query_converter_test.go
Comment (linasm) [inline src/query/storage/promremote/query_converter_test.go:151]: This is a confusing formatting.

**Review by linasm — APPROVED** [review]
LGTM with a few comments.

File: src/query/storage/promremote/options.go
Comment (soundvibe) [inline src/query/storage/promremote/options.go:48]: nit: I think it would be more clear if `EndpointOptions` would be constructed and assigned to `endpoints[i]` once.

File: src/query/server/query_test.go
Comment (Antanukas) [reply src/query/server/query_test.go:661]: linter. if I touch some old code it forces me to use WithContext. I could add `// nolint`.

File: src/query/server/query.go
Comment (robskillington) [reply src/query/server/query.go:361]: Did we remove the close from M3 storage?

Comment (robskillington) [reply src/query/server/query.go:361]: I see, yes this was removed 👍

File: src/query/server/query.go
Comment (robskillington) [inline src/query/server/query.go:492]: nit: This could just be `opts, err := promremote.NewOptions(cfg.PrometheusRemoteBackend, scope)` yeah? Given shadowing err again would be just fine and opts is declared only just above here.

File: src/query/server/query.go
Comment (robskillington) [reply src/query/server/query.go:502]: +1 can scope this to just the if
```go
if err := backendStorage.Close(); err != nil {
  logger.Error("error when closing storage", zap.Error(err))
}
```

File: src/query/server/query_test.go
Comment (robskillington) [inline src/query/server/query_test.go:655]: Nice refactor, this is good to have common across the tests here 👍

File: src/query/storage/promremote/options.go
Comment (robskillington) [inline src/query/storage/promremote/options.go:40]: nit: Typically we use 
```go
foo := make(bar, 0, len(from))
for _, elem := range from {
  foo = append(foo, bar{...})
}
```

Just so if in the inner loop there's a continue or a break or an early return and there's no empty elements left in the array that was built.

File: src/query/storage/promremote/promremotetest/test_server.go
Comment (robskillington) [inline src/query/storage/promremote/promremotetest/test_server.go:65]: nit: You can use `httptest` here which does a few of these things for you:
https://pkg.go.dev/net/http/httptest#NewServer

i.e. 
```go
testPromServer.svr := httptest.NewServer() // creates and starts server
testPromServer.addr = testPromServer.svr.URL

// Later to close
testPromServer.svr.Close()
```

File: src/query/storage/promremote/query_coverter.go
Comment (robskillington) [inline src/query/storage/promremote/query_coverter.go:67]: nit: Again prefer and more consistent in M3 code base to do `foo := make(bar, 0, len(expected))` and `foo = append(foo, bar{...})` to avoid code bugs that can creep in more easily with the index based approach when code is later modified with early return and/or continue/etc.

File: src/query/storage/promremote/storage.go
Comment (robskillington) [reply src/query/storage/promremote/storage.go:72]: +1 let's perhaps return an error here for the time being?

I don't think the default catch else namespace is actually all that compelling, but returning an error would be ideal (this is what we do for m3 storage).

File: src/query/storage/promremote/storage.go
Comment (robskillington) [inline src/query/storage/promremote/storage.go:75]: nit: As per effective go is better to use a continue if it doesn't match rather than using body inside the if case:
https://golang.org/doc/effective_go#if

> This is an example of a common situation where code must guard against a sequence of error conditions. The code reads well if the successful flow of control runs down the page, eliminating error cases as they arise. Since error cases tend to end in return statements, the resulting code needs no else statements.

So this could be:
```go
		if endpoint.resolution != query.Attributes().Resolution ||
			endpoint.retention == query.Attributes().Retention {
			continue
		}

		metrics := p.endpointMetrics[endpoint.name]
		wg.Add(1)
		go func() {
			defer wg.Done()
			err := p.writeSingle(ctx, metrics, endpoint.address, bytes.NewBuffer(encoded))
			if err == nil {
				return
			}
			errLock.Lock()
			multiErr = multiErr.Add(err)
			errLock.Unlock()
		}()
```

File: src/query/storage/promremote/storage.go
Comment (robskillington) [inline src/query/storage/promremote/storage.go:80]: nit: Can also use `bytes.NewReader(...)` here which is slightly more scoped to the task of just reading the byte body.

**Review by robskillington — APPROVED** [review]
LGTM except (outside of nits/minor comments) I do feel relatively strongly  that we return an error when retention and resolution doesn't match any of the endpoints listed before we merge this (referencing this comment here https://github.com/m3db/m3/pull/3742/files#r713168023)

**Pull request merged**
